src/primitives/compression.h

```diff
@@ -0,0 +1,475 @@
+            "?" VControl VarInt
+                "b": Signed
+                "b": Compressed Signature
+                "b": Taproot Signature
+                "b": Compressed Txid
+                "b": Standard Hash
+                "b": Hash Type
+                "b": Sequence Standard
+                "bb": Sequence Encoding
+                "bbb": Script Type
+        }
+        for each input {
+            "B" Sequence                if (Sequence == 0)
+            "?" Block Height             if (Compressed TxId == 1)
+            "?" Block Index             if (Compressed TxId == 1)
+            "32 B" TxId                 if (Compressed TxId == 0)
+            "?" Vout
+            "?" Signature Length        if (Signed && Compressed Signature == 0)
+            "?" Signature                 if (Signed && Compressed Signature == 0)
+            "64 B" Signature             if (Signed && Compressed Signature == 1)
+            "1 B"  Hash Type             if (Standard Hash == 0)
+        }
+        for each output {
+            "?": Script Length            if (Script Type == 0)
+            "B 65|32|20|?": Script
+            "?": Amount
+        }
+    */
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        uint8_t version = 0;
+        if (this->m_nVersion < 4) version = this->m_nVersion;
+        uint8_t inputCount = 0;
+        if (this->m_vin.size() < 4) inputCount = this->m_vin.size();
+        uint8_t outputCount = 0;
+        if (this->m_vout.size() < 4) outputCount = this->m_vout.size();
+        uint8_t lockTime = 0;
+        if (this->m_shortendLockTime) lockTime = 1; else if (this->m_nLockTime != 0) lockTime = 2;
+
+        uint64_t control = version;
+        control |= inputCount << 2;
+        control |= outputCount << 4;
+        control |= lockTime << 6;
+        s << VARINT(control);
+
+        if (!version) s << VARINT(this->m_nVersion);
+        if (!inputCount) s << VARINT(this->m_vin.size());
+        if (!outputCount) s << VARINT(this->m_vout.size());
+        if (lockTime) s << VARINT(this->m_nLockTime);
+
+        for (size_t index = 0; index < std::max(this->m_vin.size(), this->m_vout.size()); index++) {
+            uint64_t vControl = 0;
+            if (this->m_vin.size() > index) vControl = this->m_vin.at(index).GetMetadata();
+            uint8_t serializedScriptType = 0;
+            if (this->m_vout.size() > index) serializedScriptType = this->m_vout.at(index).GetSerializedScriptType();
+            vControl |= serializedScriptType << 8;
+            s << VARINT(vControl);
+            if (this->m_vin.size() > index) this->m_vin.at(index).Serialize(s);
+            if (this->m_vout.size() > index) this->m_vout.at(index).Serialize(s);
+        }
+    }
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        uint64_t control = std::numeric_limits<uint64_t>::max();
+        s >> VARINT(control);
+        uint8_t version = control & 0b11;
```

Comment:
In 89b52782f8d6aca75d49048da1e137fbf615c0e6:

Fine to just use `1` everywhere rather than `0b1`. 1 is the same in every base.

Reply:
I prefer the styling to stay the same when I'm working with a unique type like bits to make it easier to read what the goal is. But I can change it if to fit standard practice.

Reply:
Oh, oops, `0b11` is fine. Though it would be a bit more idiomatic to use `3` or `0x03`.

For 1 it's *far* more idiomatic to just use 1 rather than 0b1. It's a very standard idiom to use `& 1` to get the lowest bit, `^ 1` to flip it, etc., and it's understood that if you're doing bitwise operations that you have a bitwise type.

Reply:
Fixed, Thanks!